cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -575,10 +575,13 @@ private void assertConsistentTypes(Class<? extends Mapper> firstMapperClass,
     if (mapperInputs.isEmpty()) {
       // user is not going through our APIs to add input; leave the job's input format to user and simply
       // return the mapper output types of the mapper configured on the job
-      return resolveClass(rootMapperclass, Mapper.class);
+      return resolveClass(job.getConfiguration(), MRJobConfig.MAP_CLASS_ATTR, Mapper.class);
     }
 
     Class<? extends Mapper> firstMapperClass = null;
+    // resolved mapper class is for when the configured Mapper is null, so it resolves to the Mapper class configured
+    // on the job
+    Class<? extends Mapper> firstResolvedMapperClass = null;
```

Comment:
well, the default mapper class is defined higher up on line 572:
`Class<? extends Mapper> rootMapperClass = job.getMapperClass();`

This is just the resolved mapper class for the first input (first `mapperInput.getMapper()`, but defaulting to `rootMapperClass` if the first `mapperInput.getMapper()` returns null).
We keep track of the unresolved mapper class also to return at the bottom of this method.